source/extensions/filters/network/http_connection_manager/config.cc

```diff
@@ -463,8 +463,15 @@ HttpConnectionManagerConfig::HttpConnectionManagerConfig(
   }
 
   const auto& filters = config.http_filters();
+  DependencyManager dependency_manager;
   for (int32_t i = 0; i < filters.size(); i++) {
-    processFilter(filters[i], i, "http", filter_factories_, "http", i == filters.size() - 1);
+    processFilter(filters[i], i, "http", "http", i == filters.size() - 1, filter_factories_,
```

Comment:
I'm mildly concerned about the complexity costs here.  We're doing a lot of proto copying and a lot of f * d * r looping for [filters, dependencies, routes] for something I am not sure most folks need.  cc @htuch because it makes me twitchy even if the control plane is fundamentally untrusted.

Offhand, can we avoid the proto copy since all of this under the scope of this config.  Also can we have a fast-path where we exit validDecodeDependencies() if there are no dependencies?  AFIK no Envoy filters have any, so it'd mean only folks who use the feature pay the computational cost.  Also we could skip the insert for filters which don't provide/depend for those who do annotate dependencies.

Reply:
Where is the proto copying happening? I don't see why we would have to proto copy to validate dependencies, the protos in question should all be const&. Unless there is some kind of Any unpacking occurring?

Reply:
I believe a copy will occur here: https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/network/http_connection_manager/dependency_manager.h#L28

Is the concern about the # of copy operations or the size of the operations? Many filters will have empty/default FilterDependencies, which will be copied but won't be iterated through [(here)](https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/network/http_connection_manager/dependency_manager.cc#L25). I think skipping the insert would work fine, but it means DependencyManager no longer truly knows what the filterchain is, and I'm not sure about the conceptual-simplicity vs optimization tradeoff.

Reply:
I think we should try to avoid the copy, as @htuch outlines

Reply:
This is more of a change at the DependencyManager level than at HcmConfig, so I split out a separate PR to add this cleanup there. Does this look OK? https://github.com/envoyproxy/envoy/pull/15856

Reply:
That does the bit you didn't want to do, not inserting no-op filters, but still results in copies, no?

either way I'd be inclined to just do the change in place here.

Reply:
If this is happening on a per-route basis then the copy will be materially bad. If it's happening once per filter-chain (i.e. once per listener), it's less of a big deal (since huge numbers of listeners and churn is not generally something that happens, there are other costs that make that prohibitive). But yeah, feel free to make the change here.

Reply:
Hmm, where is the copy happening? registerFilter takes a const&, the copy operation happens at `filter_chain_.push_back()` which is skipped if empty. The reason I'd rather not do the change in place is, it kind contradicts the DependencyManager API which says "register your filters in decode path order". For DependencyManager to skip empty-dependencies filters, it's an implementation detail, whereas HcmConfig would be kind of hacking the DependencyManager. No difference in the immediate future, but I worry about clarity and colocation of where these assumptions are made.

From my understanding, this is once per filter-chain factory i.e. on listener config update, but I'm not 100% confident.

Reply:
As long as it's once per filter chain factory and these are network factories, I don't mind either way.